Gauge.CSharp.Lib/ScenarioDataStore.cs

```diff
@@ -1,19 +1,23 @@
 using System.Collections.Concurrent;
-using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
 
 namespace Gauge.CSharp.Lib
 {
     public class ScenarioDataStore
     {
-        private static ThreadLocal<ConcurrentDictionary<object, object>> store = new ThreadLocal<ConcurrentDictionary<object, object>>(() => new ConcurrentDictionary<object, object>());
+        private static ThreadLocal<ConcurrentDictionary<object, object>> store = new ThreadLocal<ConcurrentDictionary<object, object>>(() => new ConcurrentDictionary<object, object>(), true);
```

Comment:
@nablik - can you please explain what was the issue caused here? I'd like to know why the ThreadLocal needs to be visible to other threads.

Reply:
Hi @sriv, in multithread execution e.g
```csharp
 public void AllDataStoreParallelInvoke()
        {
            Task[] tasks = new Task[12]
            {
                Task.Factory.StartNew(() => ScenarioDataStore.Add("sckey7", "Scenario7")),
                Task.Factory.StartNew(() => ScenarioDataStore.Add("sckey8", "Scenario8")),
                Task.Factory.StartNew(() => ScenarioDataStore.Add("sckey9", "Scenario9")),
                Task.Factory.StartNew(() => ScenarioDataStore.Add("sckey10", "Scenario10")),
                Task.Factory.StartNew(() => SpecDataStore.Add("spec1", "Spec1")),
                Task.Factory.StartNew(() => SpecDataStore.Add("spec2", "Spec2")),
                Task.Factory.StartNew(() => SpecDataStore.Add("spec3", "Spec3")),
                Task.Factory.StartNew(() => SpecDataStore.Add("spec4", "Spec4")),
                Task.Factory.StartNew(() => SuiteDataStore.Add("suite1", "Suite1")),
                Task.Factory.StartNew(() => SuiteDataStore.Add("suite2", "Suite2")),
                Task.Factory.StartNew(() => SuiteDataStore.Add("suite3", "Suite3")),
                Task.Factory.StartNew(() => SuiteDataStore.Add("suite4", "Suite4"))
            };
            Task.WaitAll(tasks);
            Assert.AreEqual(SpecDataStore.Get("spec2"), "Spec2");
            Assert.AreEqual(SpecDataStore.Get("spec3"), "Spec3");
            Assert.AreEqual(SpecDataStore.Get("spec4"), "Spec4");
            Assert.AreEqual(SuiteDataStore.Get("suite1"), "Suite1");
            Assert.AreEqual(SuiteDataStore.Get("suite2"), "Suite2");
            Assert.AreEqual(SuiteDataStore.Get("suite3"), "Suite3");
            Assert.AreEqual(SuiteDataStore.Get("suite4"), "Suite4");
            Assert.AreEqual(ScenarioDataStore.Get("sckey7"), "Scenario7");
            Assert.AreEqual(ScenarioDataStore.Get("sckey8"), "Scenario8");
            Assert.AreEqual(ScenarioDataStore.Get("sckey9"), "Scenario9");
            Assert.AreEqual(ScenarioDataStore.Get("sckey10"), "Scenario10");
        }
```

the data is stored in ThreadLocal.Values instead of ThreadLocal.Value (even though the code is explicitly calling  store.Value)
![image](https://user-images.githubusercontent.com/11925279/95970506-f309d580-0e07-11eb-96bb-aa13f8905bb5.png)
and Values list actually contain all the data
![image](https://user-images.githubusercontent.com/11925279/95970624-159bee80-0e08-11eb-900d-f0ec86b6bc02.png)
so to have access to the Values, the values must be visible to the other thread.
I just discovered with that unit test that the new solution, unfortunately, does not work either in parrallel execution.
so thye code changes are there to make it happening.

Reply:
Thanks for this, I'll look into this a bit deeper. I am in middle of something, so this may take a few days.

Reply:
Thanks.

Reply:
I spent some time on this. I feel that the tests do not represent the actual scenario. Here's why:

The expectation is that a `*Datastore.Add` and `*DataStore.Get` should be local to a thread, because when gauge-dotnet runs the specs, it will run [one stream per thread](https://github.com/getgauge/gauge-dotnet/blob/master/src/Executor/ExecutorPool.cs#L23). Hence, `ScenarioDataStore.Add` and `ScenarioDataStore.Get` must be called in the same thread. 

This was done to ensure that one stream does not access datastore from another stream. Setting `trackAllValues=true` in `ThreadLocal` contradicts this expectation, IMO.

If you were to refactor the tests to ensure that datastore writes and reads are done in the same thread, but use multiple threads for each write/read combination, you may find that the tests pass (and represent the use case). `Task.Factory.StartNew` creates a new thread for each task by default.

You could either use a custom task scheduler, or use task continuation for assertion with `ExecuteSynchrounously` as [`TaskContinuationOptions`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcontinuationoptions?view=netcore-3.1)

Also, consider using [`Task.Run` instead of `Task.Factory.StartNew`](https://devblogs.microsoft.com/pfxteam/task-run-vs-task-factory-startnew/) if you do not plan to use a custom TaskScheduler.

If I revert your changes for `trackAllValues=true` and refactor one test to be like:

```
        [Test]
        public void AllDataStoreParallelInvoke()
        {
            Task.WaitAll(new Task[12]
            {
                Task.Run(() => ScenarioDataStore.Add("sckey7", "Scenario7"))
                    .ContinueWith((o) => Assert.AreEqual("Scenario7", ScenarioDataStore.Get("sckey7")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => ScenarioDataStore.Add("sckey8", "Scenario8"))
                    .ContinueWith((o) => Assert.AreEqual("Scenario8", ScenarioDataStore.Get("sckey8")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => ScenarioDataStore.Add("sckey9", "Scenario9"))
                    .ContinueWith((o) => Assert.AreEqual("Scenario9", ScenarioDataStore.Get("sckey9")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => ScenarioDataStore.Add("sckey10", "Scenario10"))
                    .ContinueWith((o) => Assert.AreEqual("Scenario10", ScenarioDataStore.Get("sckey10")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => SpecDataStore.Add("spec1", "Spec1"))
                    .ContinueWith((o) => Assert.AreEqual("Spec1", SpecDataStore.Get("spec1")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => SpecDataStore.Add("spec2", "Spec2"))
                    .ContinueWith((o) => Assert.AreEqual("Spec2", SpecDataStore.Get("spec2")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => SpecDataStore.Add("spec3", "Spec3"))
                    .ContinueWith((o) => Assert.AreEqual("Spec3", SpecDataStore.Get("spec3")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => SpecDataStore.Add("spec4", "Spec4"))
                    .ContinueWith((o) => Assert.AreEqual("Spec4", SpecDataStore.Get("spec4")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => SuiteDataStore.Add("suite1", "Suite1"))
                    .ContinueWith((o) => Assert.AreEqual("Suite1", SuiteDataStore.Get("suite1")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => SuiteDataStore.Add("suite2", "Suite2"))
                    .ContinueWith((o) => Assert.AreEqual("Suite2", SuiteDataStore.Get("suite2")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => SuiteDataStore.Add("suite3", "Suite3"))
                    .ContinueWith((o) => Assert.AreEqual("Suite3", SuiteDataStore.Get("suite3")), TaskContinuationOptions.ExecuteSynchronously),
                Task.Run(() => SuiteDataStore.Add("suite4", "Suite4"))
                    .ContinueWith((o) => Assert.AreEqual("Suite4", SuiteDataStore.Get("suite4")), TaskContinuationOptions.ExecuteSynchronously),
            });
        }
```

This test passes. Let me know what you think.

Reply:
I am going to apply those changes and update my PR.